python/python/lance/arrow.py

```diff
@@ -11,14 +11,444 @@
+            typ = FixedShapeImageTensorType(value_type, shape)
+            return pa.ExtensionArray.from_storage(typ, images.storage)
+        elif isinstance(
+            images, (ImageURIArray, EncodedImageArray, FixedShapeImageTensorArray)
+        ):
+            return images
+        elif isinstance(images, (list, tuple, Iterable)):
+            return pa.ExtensionArray.from_storage(
+                ImageURIType(), pa.array(images, type=pa.string())
+            )
+
+        else:
+            raise TypeError("Cannot build a ImageArray from {}".format(type(images)))
+
+
+class ImageURIArray(ImageArray):
+    """
+    Array of image URIs. URIs may represent local files or remote files on the web,
+    S3 or GCS if they are accessible by the machine executing this.
+    """
+
+    @classmethod
+    def from_uris(cls, uris: Union[pa.StringArray, Iterable[Union[str, Path]]]):
+        """
+        Create an ImageURIArray from a pa.StringArray or an iterable.
+
+        Parameters
+        ----------
+        uris : Union[pa.StringArray, Iterable[Union[str, Path]]]
+
+        Returns
+        -------
+        ImageURIArray
+            Array of image URIs
+
+        Examples
+        --------
+        >>> uris = ["file::///tmp/1.png"]
+        >>> ImageURIArray.from_uris(uris)
+        <lance.arrow.ImageURIArray object at 0x...>
+        ['file::///tmp/1.png']
+        """
+
+        if not isinstance(uris, pa.StringArray):
+            uris = pa.array((str(uri) for uri in uris), type=pa.string())
+        else:
+            raise TypeError("Cannot build a ImageURIArray from {}".format(type(uris)))
+
+        return cls.from_storage(ImageURIType(), uris)
+
+    def read_uris(self):
+        """
+        Read the images from the URIs into memory and return an EncodedImageArray
+
+        Returns
+        -------
+        EncodedImageArray
+            Array of encoded images
+
+        Examples
+        --------
+        >>> import os
+        >>> uris = [os.path.join(os.path.dirname(__file__), "../tests/images/1.png")]
+        >>> uri_array = ImageURIArray.from_uris(uris)
+        >>> uri_array.read_uris()
+        <lance.arrow.EncodedImageArray object at 0x...>
+        ...
+        """
+        from urllib.error import URLError
+        from urllib.parse import urlparse
+        from urllib.request import Request, urlopen
+
+        from pyarrow import fs
+
+        def download(url):
+            req = Request(url)
+            try:
+                return urlopen(req).read()
+            except URLError as e:
+                if hasattr(e, "reason"):
+                    print("Failed to reach the server: ", e.reason)
+                elif hasattr(e, "code"):
+                    print(
+                        "The server could not fulfill the request. Error code: ", e.code
+                    )
+
+        allowed_schemes = ["s3", "gs", "file", ""]
+        if os.name == "nt":
+            allowed_schemes.append("d")
```

Comment:
I'm a little bit unsure about this.

Reply:
Can we just let `pyarrow.fs.FileSystem` to handle what is acceptable schema for us?

So your for loop can be

```
for uri in self.storage:
    parsed_uri = urlparse(uri.as_py())
    if parsed_uri.schema in ("http", "https"):
         image.append(download(url))
   else:
         filesystme, path = fs.FileSystme.from_uri(uri.as_py())
```

Reply:
Yeah that's easier for sure :D. Changed.